src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MessagePackSerializerContext.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Composition
+{
+    using System.Collections.Concurrent;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Reflection;
+    using MessagePack;
+    using MessagePack.Formatters;
+    using MessagePack.Resolvers;
+
+#pragma warning disable CS8602 // possible dereference of null reference
+#pragma warning disable CS8604 // null reference as argument
+
+    internal class MessagePackSerializerContext : MessagePackSerializerOptions, IDisposable
+    {
+        public Resolver CompositionResolver { get; }
+
+        private static IFormatterResolver GetIFormatterResolver(IFormatterResolver resolver) => CompositeResolver.Create(
+                [
+                    new IgnoreFormatter<System.Threading.CancellationToken>(),
+                    new IgnoreFormatter<System.Threading.Tasks.Task>(),
+                ],
+                [resolver]);
+
+        public MessagePackSerializerContext(IFormatterResolver resolver, Resolver compositionResolver)
+            : base(GetIFormatterResolver(resolver))
+        {
+            this.deserializingObjectTable = new ConcurrentDictionary<uint, object?>();
+            this.CompositionResolver = compositionResolver;
+        }
+
+        public MessagePackSerializerContext(int estimatedObjectCount, IFormatterResolver resolver, Resolver compositionResolver)
+            : base(GetIFormatterResolver(resolver))
+        {
+            this.serializingObjectTable = new ConcurrentDictionary<object, uint>(SmartInterningEqualityComparer.Default);
+            this.CompositionResolver = compositionResolver;
+        }
+
+        private ConcurrentDictionary<object, uint>? serializingObjectTable;
```

Comment:
No reason to use `ConcurrentDictionary` because serialization is never multi-threaded. And each distinct serialization job must necessarily have its own instance of this class.